### DISCUSSION TITLE
KEM algorithms encap and decap with oqsprovider

### DISCUSSION THREAD

**Discussion by NarjasHadjer** [open-quantum-safe/oqs-provider#484] [Q&A]
Hi,
I'm attempting to encapsulate a private key generated with Kyber512 on my Raspberry Pi 3B+ using Pi OS, but I'm encountering the error: "Operation not supported for this key type."
Could you advise on how to resolve this issue?
![received_1004066174796481](https://github.com/user-attachments/assets/dd4cb6b4-f1bd-411b-bea3-93619f757211)
![received_26175712578738820](https://github.com/user-attachments/assets/6b09d82e-ada9-4c28-a56e-7327e8a19d67)

Thanks.

**Comment by baentsch** [comment]
I assume you have built `oqsprovider` with [OQS_KEM_ENCODERS](https://github.com/open-quantum-safe/oqs-provider/blob/main/CONFIGURE.md#oqs_kem_encoders)? If it then still doesn't work, I'm afraid you're out of luck: The openssl `pkeyutil` app has not been tested with `oqsprovider` and I'm not sure it's fully wired up for provider support. Maybe a question worth while asking in the OpenSSL forum? If you are convinced it should work, please consider contributing a PR to the `oqsprovider` test scripts.

**Reply by NarjasHadjer** [reply]
Effectively, i have built oqsprovider with OQS_KEM_ENCODERS,I am unable to either encrypt a text file or encapsulate the secret key. I am trying to apply OpenSSL commands and adapt them for PQC algorithms, but it's not working. When using RSA, for example, I am able to encrypt data, encapsulate, and decapsulate a secret key. However, when I try to do the same with Kyber512 by simply changing the algorithm name as shown in the screenshots, it does not work. I have also, upgraded OpenSSL to version 3.3.0, but i get the same errors. I have also noticed that with OQS_KEM_ENCODERS enabled, I encounter errors with DS algorithms, even though I tested the digital signature algorithms normally before enabling it.

**Comment by beldmit** [comment]
You want to encapsulate, not encrypt. Encapsulation was recently added by me to the OpenSSL master branch and will be a part of the upcoming 3.4 release

**Reply by NarjasHadjer** [reply]
Could you Sir please check if there is any error in the syntax of my command? I am unable to either encrypt a text file or encapsulate the secret key. I am trying to apply OpenSSL commands and adapt them for PQC algorithms, but it's not working. Could you kindly correct the encryption command, as the encapsulation is currently not supported?

**Reply by beldmit** [reply]
https://github.com/openssl/openssl/pull/25127 is the PR, you can cherry-pick it, rebuild openssl, and use the syntax from the tests

#Encap/decap tests
$ openssl pkeyutl -encap -pubin -inkey rsa_pub.pem -secret secret.bin -out encap_out.bin
$ openssl pkeyutl -decap -inkey rsa_priv.pem -in encap_out.bin -out decap_out.bin
// decap_out is equal to secret

**Reply by NarjasHadjer** [reply]
I now understand what you meant in your previous comment, sir. Indeed, I know that I need to use the encap option instead of encrypt, but the encap and decap options are not available with the openssl pkeyutl command. I thought this might have been a change in OpenSSL 3.3.0.
![received_3987804438209844](https://github.com/user-attachments/assets/81a175c5-84d7-444f-b02e-1061ba0e8b86)
![received_8402657913106118](https://github.com/user-attachments/assets/574c0706-1bb5-4fc8-930d-08045db7426e)

**Reply by baentsch** [reply]
Thanks for pointing out the new/3.4.0 encap/decap feature, @beldmit . Worth while adding a comment at https://github.com/open-quantum-safe/oqs-provider?tab=readme-ov-file#note-on-openssl-versions and maybe a (openssl version-dependent) test in https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/runtests.sh?

**Reply by beldmit** [reply]
@NarjasHadjer to get this feature here and now, you'll have to apply code from the PR to openssl code (or wait till, I think, October 2024 for the 3.4 release). 

@baentsch yes, it probably is worth a test and mentioning.
